pkg/ovs/ovsctl/interface.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Antrea Authors
+//
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovsctl
+
+import (
+	"net"
+	"os/exec"
+)
+
+// TracingRequest defines tracing request parameters.
+type TracingRequest struct {
+	InPort string // Input port.
+	SrcIP  net.IP
+	DstIP  net.IP
+	SrcMAC net.HardwareAddr
+	DstMAC net.HardwareAddr
+	Flow   string
+	// Whether in_port field in Flow can override InPort.
+	AllowOverrideInPort bool
+}
+
+// OVSCtlClient is an interface for executing OVS "ovs-ofctl" and "ovs-appctl"
+// commands.
+type OVSCtlClient interface {
+	// DumpFlows returns flows of the bridge.
+	DumpFlows(args ...string) ([]string, error)
+	// DumpMatchedFlows returns the flow which exactly matches the matchStr.
+	DumpMatchedFlow(matchStr string) (string, error)
+	// DumpTableFlows returns all flows in the table.
+	DumpTableFlows(table uint8) ([]string, error)
+	// DumpGroups returns OpenFlow groups of the bridge.
+	DumpGroups(args ...string) ([][]string, error)
+	// RunOfctlCmd executes "ovs-ofctl" command and returns the outputs.
+	RunOfctlCmd(cmd string, args ...string) ([]byte, error)
+	// Trace executes "ovs-appctl ofproto/trace" to perform OVS packet tracing.
+	Trace(req *TracingRequest) (string, error)
+}
+
+type ovsCtlClient struct {
+	bridge string
+}
+
+func NewClient(bridge string) OVSCtlClient {
+	return &ovsCtlClient{bridge}
+}
```

Comment:
I think normally the concrete type is not part of interface and should be in "pkg/ovs/ovsctl/ofctl.go" in this case.

And a more extensive way is to return the concrete type, not the interface so that you can add more methods to this type and clients can receive it as other interfaces.
https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Reply:
Ok. I moved ovsCtlClient to ofctl.go, and changed NewClient() to return *ovsCtlClient.